components/raftstore/src/store/entry_storage.rs

```diff
@@ -530,6 +551,76 @@ pub fn init_applied_term<ER: RaftEngine>(
     }
 }
 
+/// When a peer(follower) receives a TransferLeaderMsg, it enters the
+/// CacheWarmupState. When the peer becomes leader or it doesn't
+/// become leader before a deadline, it exits the state.
+#[derive(Clone, Debug)]
+pub struct CacheWarmupState {
+    range: (u64, u64),
+    is_task_timeout: bool,
+    is_stale: bool,
+    started_at: Instant,
+}
+
+impl CacheWarmupState {
+    pub fn new() -> Self {
+        CacheWarmupState::new_with_range(INVALID_INDEX, INVALID_INDEX)
+    }
+
+    pub fn new_with_range(low: u64, high: u64) -> Self {
+        CacheWarmupState {
+            range: (low, high),
+            is_task_timeout: false,
+            is_stale: false,
+            started_at: Instant::now(),
+        }
+    }
+
+    pub fn range(&self) -> (u64, u64) {
+        self.range
+    }
+
+    /// How long has it been in this state.
+    pub fn elapsed(&self) -> Duration {
+        self.started_at.saturating_elapsed()
+    }
+
+    /// Whether the warmup task is already timeout.
+    pub fn is_task_timeout(&self) -> bool {
+        self.is_task_timeout
+    }
+
+    /// Check whether the task is timeout.
+    pub fn check_task_timeout(&mut self, duration: Duration) -> bool {
```

Comment:
We should hardcode the duration here in this function. Otherwise you cannot defend code patterns that use different durations and cause inconsistency.

Reply:
Do you mean to store the duration of timeout in the struct like the following?

```rust
pub struct CacheWarmupState {
    ...
    task_deadline: Duration,
    is_task_timeout: bool,
}

impl CacheWarmupState {
    fn check_task_timeout(&mut self) {
        if self.is_task_timeout {
            return true;
        }
        if self.elapsed() > self.task_deadline {
            WARM_UP_ENTRY_CACHE_COUNTER.timeout.inc();
            self.is_task_timeout = true;
        }
        self.is_task_timeout
    }
}
```

Reply:
Oh, I was mainly thinking about `check_stale`. How about this:
```
struct State {
  cached_elapsed_duration: Duration
}

impl State {
  fn check_timeout(&mut self, timeout: Duration) {
    if self.cached_elapsed_duration > timeout { return true; }
    self.cached_elapsed_duration = self.start.elapsed();
    self.cached_elapsed_duration > timeout
  }
}
```

Reply:
>  I was mainly thinking about `check_stale`

For `check_stale`, I personally prefer the following solution to check if the state is stale or not  
```rust
impl State {
    pub fn is_stale(&self) -> bool {
        self.elapsed() >= MAX_WARMED_UP_CACHE_KEEP_TIME
    }
}
```

Reply:
> Oh, I was mainly thinking about check_stale.

>We should hardcode the duration here in this function. Otherwise you cannot defend code patterns that use different durations and cause inconsistency.

Considering the two comments above, I have another idea
```rust
struct State {
    is_stale: bool,
}
impl State {
  fn check_stale(&mut self) {
        if self.is_stale {
            return true;
        }
        if self.elapsed() > MAX_WARMED_UP_CACHE_KEEP_TIME {
            self.is_stale = true;
        }
        self.is_stale
  }
}
```

Reply:
I'm fine with it, but `check_task_timeout` still bothers me. My previous comment seems to fix both functions at the same time.